InnerEye/ML/SSL/datamodules_and_datasets/dataset_cls_utils.py

```diff
@@ -0,0 +1,32 @@
+#  ------------------------------------------------------------------------------------------
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+#  ------------------------------------------------------------------------------------------
+from typing import Any, Tuple, Union
+
+import torch
+
+OptionalIndexInputAndLabel = Union[Tuple[torch.Tensor, int], Tuple[int, torch.Tensor, int]]
+
+
+class InnerEyeDataClassBaseWithReturnIndex:
+    """
+    Class to be use with double inheritance with a VisionDataset.
+    Overloads the __getitem__ function so that we can optionally also return
+    the index within the dataset.
+    """
+
+    def __init__(self, root: str, return_index: bool, **kwargs: Any) -> None:
+        self.return_index = return_index
+        super().__init__(root=root, **kwargs)  # type: ignore
+
+    def __getitem__(self, index: int) -> OptionalIndexInputAndLabel:
+        img, target = super().__getitem__(index)  # type: ignore
```

Comment:
maybe treat this as a tuple in case user wants to return multiple objects with a dataloader - you can just append index in the front e.g. `index, *item`

Reply:
good point, updated